crates/provider/src/fillers/nonce.rs

```diff
@@ -0,0 +1,163 @@
+use crate::{
+    fillers::{FillerControlFlow, TxFiller},
+    provider::SendableTx,
+    Provider,
+};
+use alloy_network::{Network, TransactionBuilder};
+use alloy_primitives::Address;
+use alloy_transport::{Transport, TransportResult};
+use dashmap::DashMap;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+/// A [`TxFiller`] that fills nonces on transactions.
+///
+/// The filler will fetch the transaction count for any new account it sees,
+/// store it locally and increment the locally stored nonce as transactions are
+/// sent via [`Provider::send_transaction`].
+///
+/// # Note
+///
+/// - If the transaction request does not have a sender set, this layer will
+///  not fill nonces.
+/// - Using two providers with their own nonce layer can potentially fill
+///  invalid nonces if transactions are sent from the same address, as the next
+///  nonce to be used is cached internally in the layer.
+///
+/// # Example
+///
+/// ```
+/// # use alloy_network::{NetworkSigner, EthereumSigner, Ethereum};
+/// # use alloy_rpc_types::TransactionRequest;
+/// # use alloy_provider::{ProviderBuilder, RootProvider, Provider};
+/// # async fn test<S: NetworkSigner<Ethereum> + Clone>(url: url::Url, signer: S) -> Result<(), Box<dyn std::error::Error>> {
+/// let provider = ProviderBuilder::new()
+///     .with_nonce_management()
+///     .signer(signer)
+///     .on_http(url)?;
+///
+/// provider.send_transaction(TransactionRequest::default()).await;
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Debug, Clone, Default)]
+pub struct NonceFiller {
+    nonces: DashMap<Address, Arc<Mutex<Option<u64>>>>,
```

Comment:
this is from @onbjerg mostly. I think the `Arc<Mutex<Option` could be simplified to `Option<AtomicU64>>`. I have also not checked for the race conditions that we had to mitigate in ethers

Reply:
the issue w just `Option<AtomicU64>` is that you might have multiple tx's in flight and they would both try to init the nonce, so we need to lock it afaik. it's only because of init, otherwise the lock is pretty much useless

Reply:
ya my considered take would be OnceLock<AtomicU64>